src/hotspot/os/posix/perfMemory_posix.cpp

```diff
@@ -578,52 +589,41 @@ static char* get_user_name_slow(int vmid, int nspid, TRAPS) {
       continue;
     }
 
-    struct dirent* udentry;
     errno = 0;
-    while ((udentry = os::readdir(subdirp)) != nullptr) {
-
-      if (filename_to_pid(udentry->d_name) == searchpid) {
-        struct stat statbuf;
-        int result;
-
-        char* filename = NEW_C_HEAP_ARRAY(char,
-                                          strlen(usrdir_name) + strlen(udentry->d_name) + 2,
-                                          mtInternal);
-
-        strcpy(filename, usrdir_name);
-        strcat(filename, "/");
-        strcat(filename, udentry->d_name);
-
-        // don't follow symbolic links for the file
-        RESTARTABLE(::lstat(filename, &statbuf), result);
-        if (result == OS_ERR) {
-           FREE_C_HEAP_ARRAY(char, filename);
-           continue;
-        }
-
-        // skip over files that are not regular files.
-        if (!S_ISREG(statbuf.st_mode)) {
-          FREE_C_HEAP_ARRAY(char, filename);
-          continue;
-        }
-
-        // compare and save filename with latest creation time
-        if (statbuf.st_size > 0 && statbuf.st_ctime > oldest_ctime) {
-
-          if (statbuf.st_ctime > oldest_ctime) {
-            char* user = strchr(dentry->d_name, '_') + 1;
+    // the filename corresponding to the vmid (or nspid)
+    const char* filename = get_sharedmem_filename(usrdir_name, vmid, nspid);
+    struct stat statbuf;
+    int result;
+    // check if it exists; don't follow symbolic links for the file
+    RESTARTABLE(::lstat(filename, &statbuf), result);
+    if (result == OS_ERR) {
+      FREE_C_HEAP_ARRAY(char, filename);
+      FREE_C_HEAP_ARRAY(char, usrdir_name);
+      os::closedir(subdirp);
+      continue;
+    }
+    // skip over files that are not regular files.
+    if (!S_ISREG(statbuf.st_mode)) {
+      FREE_C_HEAP_ARRAY(char, filename);
+      FREE_C_HEAP_ARRAY(char, usrdir_name);
+      os::closedir(subdirp);
+      continue;
+    }
+    FREE_C_HEAP_ARRAY(char, filename);
```

Comment:
If you move this to immediately after lstat then you don't need it in the if-block

Reply:
Thank you David for these inputs; they were useful. I've updated the PR to use those suggestions. Locally tests continue to pass. I've triggered CI testing with these new changes.